src/main/java/com/cleanroommc/groovyscript/compat/inworldcrafting/PistonPush.java

```diff
@@ -0,0 +1,161 @@
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class PistonPush extends VirtualizedRegistry<PistonPush.Recipe> {
+
+    private final List<Recipe> recipes = new ArrayList<>();
+
+    @Optional.Method(modid = "jei")
+    @GroovyBlacklist
+    public List<PistonPushRecipeCategory.RecipeWrapper> getRecipeWrappers() {
+        return this.recipes.stream().map(PistonPushRecipeCategory.RecipeWrapper::new).collect(Collectors.toList());
+    }
+
+    @Override
+    public void onReload() {
+        this.recipes.addAll(getBackupRecipes());
+        getScriptedRecipes().forEach(this.recipes::remove);
+    }
+
+    public void add(Recipe recipe) {
+        this.recipes.add(recipe);
+        addScripted(recipe);
+    }
+
+    public boolean remove(Recipe recipe) {
+        if (this.recipes.remove(recipe)) {
+            addBackup(recipe);
+            return true;
+        }
+        return false;
+    }
+
+    public RecipeBuilder recipeBuilder() {
+        return new RecipeBuilder();
+    }
+
+    public static class Recipe {
+
+        private final IIngredient input;
+        private final ItemStack output;
+        private final int maxConversionsPerPush;
+        private final int minHarvestLevel;
+        private final Closure<Boolean> beforeRecipe;
+
+        public Recipe(IIngredient input, ItemStack output, int maxConversionsPerPush, int minHarvestLevel, Closure<Boolean> beforeRecipe) {
+            this.input = input;
+            this.output = output;
+            this.maxConversionsPerPush = maxConversionsPerPush;
+            this.minHarvestLevel = minHarvestLevel;
+            this.beforeRecipe = beforeRecipe;
+        }
+
+        public IIngredient getInput() {
+            return input;
+        }
+
+        public ItemStack getOutput() {
+            return output;
+        }
+
+        public int getMaxConversionsPerPush() {
+            return maxConversionsPerPush;
+        }
+
+        public int getMinHarvestLevel() {
+            return minHarvestLevel;
+        }
+
+        private boolean tryRecipe(Consumer<EntityItem> entitySpawner, EntityItem entityItem, ItemStack itemStack, IBlockState pushingAgainst) {
+            if (!this.input.test(itemStack)) return false;
+            if (this.beforeRecipe != null && !ClosureHelper.call(true, this.beforeRecipe, entityItem, itemStack, pushingAgainst)) return false;
+            if (this.minHarvestLevel >= 0 && this.minHarvestLevel > pushingAgainst.getBlock().getHarvestLevel(pushingAgainst)) return false;
+            ItemStack newStack = this.output.copy();
+            if (this.maxConversionsPerPush < itemStack.getCount()) {
+                itemStack.shrink(this.maxConversionsPerPush);
+                newStack.setCount(this.maxConversionsPerPush);
+                entityItem.setItem(itemStack);
+                entitySpawner.accept(new EntityItem(entityItem.world, entityItem.posX, entityItem.posY, entityItem.posZ, newStack));
```

Comment:
`setDefaultPickupDelay` + previous comment on positions

Reply:
entity pos is not block pos, so its fine